File: Data Structures & Algorithms/minimum-window-with-characters/submission-3.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        if (t.empty() || s.length() < t.length()) return "";

        unordered_map<char, int> need;
        for (char c : t) {
            need[c]++;
        }

        unordered_map<char, int> window;
        int have = 0;
        int required = need.size();

        int resLen = INT_MAX;
        int resLeft = 0;

        int left = 0;
        for (int right = 0; right < s.length(); right++) {
            char c = s[right];
            window[c]++;

            if (need.count(c) && window[c] == need[c]) {
                have++;
            }

            while (have == required) {
                if (right - left + 1 < resLen) {
                    resLen = right - left + 1;
                    resLeft = left;
                }

                char leftChar = s[left];
                window[leftChar]--;
                if (need.count(leftChar) && window[leftChar] < need[leftChar]) {
                    have--;
                }
                left++;
            }
        }

        return resLen == INT_MAX ? "" : s.substr(resLeft, resLen);
    }
};

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-3.go
func characterReplacement(s string, k int) int {
    var count [26]int
    maxFreq := 0
    left := 0
    maxLen := 0

    for right := 0; right < len(s); right++ {
        idx := s[right] - 'A'
        count[idx]++
        if count[idx] > maxFreq {
            maxFreq = count[idx]
        }

        for (right-left+1)-maxFreq > k {
            count[s[left]-'A']--
            left++
        }

        if right-left+1 > maxLen {
            maxLen = right - left + 1
        }
    }

    return maxLen
}